### PR TITLE
Avoids infinite video decoder seek loop

### PR DESCRIPTION
**Pull request by JanuszL**
- when the video decoder is not able to start decoding from a given keyframe
  it seeks a preceding one compared to it. When it happens for the keyframe
  0 we end up in an infinite try and fail loop. This PR makes this case a hard error.
- adds graceful shutdown of the decoder when reading packets or submitting
  them to the decoder fails

## Category:
**Other** (*e.g. Documentation, Tests, Configuration*)

## Description:
- when the video decoder is not able to start decoding from a given keyframe
  it seeks a preceding one compared to it. When it happens for the keyframe
  0 we end up in an infinite try and fail loop. This PR makes this case a hard error.
- adds graceful shutdown of the decoder when reading packets or submitting
  them to the decoder fails

### Affected modules and functionalities:
- video reader

### Key points relevant for the review:
- NA

### Tests:

- [ ] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [x] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [x] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by JanuszL** [comment]
!build

**Review by awolant — APPROVED** [review]

**Comment by JanuszL** [comment]
!build

File: dali/operators/reader/loader/video_loader.cc [hunk 1]
```diff
@@ -586,11 +591,21 @@ void VideoLoader::read_file() {
...
+        ((key && last_key_frame != frame && last_key_frame != 0) ||
+          frame > last_key_frame + kStartupFrameThreshold) &&
         dec_status == VidReqStatus::REQ_NOT_STARTED) {
+        if (last_key_frame <= 0) {
+          if (vid_decoder_) {
+            vid_decoder_->finish();
+          }
+          DALI_FAIL("Decoding not started when starting from frame 0, "
+                    "no point in seeking to preceding keyframe which is "
+                    "already 0");
```
Comment (szkarpinski) [inline dali/operators/reader/loader/video_loader.cc:606]: Is there particular kind of video file or parameter set that triggers this problem? Maybe it's worth mentioning in the message? The current one doesn't seem very helpful for the user as it only refers to the internals of the decoder

Comment (JanuszL) [reply dali/operators/reader/loader/video_loader.cc:606]: The faulty sample was provided by the user in https://github.com/NVIDIA/DALI/issues/5155#issuecomment-1837082791, but I don't know how to create such one on our own and I don't think we can add this one to DALI_extra.

File: dali/operators/reader/loader/video_loader.cc [hunk 2]
```diff
@@ -539,7 +539,12 @@ void VideoLoader::read_file() {
 
 
   auto& file = get_or_open_file(req.filename);
-  DALI_ENFORCE(!file.empty(), "Cannot open video file");
+  if (file.empty()) {
+    if (vid_decoder_) {
+      vid_decoder_->finish();
```
[the remark this answers is not in the document]
Comment (szkarpinski) [reply dali/operators/reader/loader/video_loader.cc:544]: I see, thank you

**Review by szkarpinski — APPROVED** [review]

**Comment by JanuszL** [comment]
!build

**Pull request merged**